File: example-crs-webservice/crs-userspace/libs/libMSA/python/libmsa/thread/equeue.py
import queue
import threading
from typing import Any, Optional


class EventfulQueue:
    def __init__(self) -> None:
        self.queue: queue.Queue[Any] = queue.Queue()
        self.event = threading.Event()
        self.lock = threading.Lock()

    def enqueue(self, data: Any) -> None:
        with self.lock:
            self.queue.put(data)
            self.event.set()

    def dequeue(self) -> Optional[Any]:
        with self.lock:
            try:
                return self.queue.get(block=False)
            except queue.Empty:
                return None

    def task_done(self) -> None:
        with self.lock:
            if self.queue.empty():
                self.event.clear()

    def get_event(self) -> threading.Event:
        return self.event

File: example-crs-webservice/crs-userspace/libs/libMSA/go/thread/equeue.go
package thread

import (
	"container/list"
	"sync"
	//"fmt"
)

type EventfulQueue[T any] struct {
	queue *list.List
	lock  sync.Mutex
	cond  *sync.Cond
}

func NewEventfulQueue[T any]() *EventfulQueue[T] {
	q := &EventfulQueue[T]{queue: list.New()}
	q.cond = sync.NewCond(&q.lock)
	return q
}

func (eq *EventfulQueue[T]) Enqueue(data T) {
	eq.lock.Lock()
	eq.queue.PushBack(data)
	eq.cond.Broadcast()
	eq.lock.Unlock()
}

func (eq *EventfulQueue[T]) Dequeue() *T {
	eq.lock.Lock()
	defer eq.lock.Unlock()

	if eq.queue.Len() > 0 {
		elem := eq.queue.Front()
		eq.queue.Remove(elem)
		data := elem.Value.(T)
		return &data
	}
	return nil
}

func (eq *EventfulQueue[T]) GetLock() *sync.Mutex {
	return &eq.lock
}

func (eq *EventfulQueue[T]) GetCond() *sync.Cond {
	return eq.cond
}